File: Hard/85. Maximal Rectangle.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        final int R = matrix.length, C = matrix[0].length;
        final var heights = new int[C + 1];
        int maxArea = 0;
        for (var row : matrix) {
            for (int c = 0; c < C; c++) {
                heights[c] = (row[c] == '1') ? heights[c] + 1 : 0;
            }

            var stack = new ArrayDeque<Integer>();
            for (int c = 0; c <= C; c++) {
                while (!stack.isEmpty() && heights[c] < heights[stack.peekLast()]) {
                    int h = heights[stack.pollLast()];
                    int w = stack.isEmpty() ? c : c - stack.peekLast() - 1;
                    maxArea = Math.max(maxArea, h * w);
                }
                stack.offerLast(c);
            }
        }
        return maxArea;
    }
}

File: Hard/85. Maximal Rectangle.py
from typing import List


class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0
        R, C = len(matrix), len(matrix[0])
        for r in range(R):
            for c in range(C):
                matrix[r][c] = int(matrix[r][c]) # type: ignore
        for r in range(R):
            for c in range(1, C):
                if matrix[r][c] == 1:
                    matrix[r][c] += matrix[r][c - 1]
        result = 0
        for c in range(C):
            for r in range(R):
                width = matrix[r][c]
                if width == 0:
                    continue
                curr_width = width
                for k in range(r, R):
                    if matrix[k][c] == 0:
                        break
                    curr_width = min(curr_width, matrix[k][c])
                    result = max(result, curr_width * (k - r + 1))
        return result # type: ignore